scene/2d/tile_map.cpp

```diff
@@ -373,60 +612,127 @@ void TileMapLayer::_rendering_draw_quadrant_debug(TileMapQuadrant *p_quadrant) {
 
-	Transform2D gl_transform = tile_map_node->get_global_transform();
+	// Check if we should cleanup everything.
+	bool forced_cleanup = in_destructor || !enabled || !tile_map_node->is_inside_tree() || !tile_set.is_valid() || !tile_map_node->is_visible_in_tree();
+	if (forced_cleanup) {
+		// Clean everything.
+		for (KeyValue<Vector2i, CellData> &kv : tile_map) {
+			_physics_clear_cell(kv.value);
+		}
+	} else {
+		if (_physics_was_cleaned_up || dirty.flags[DIRTY_FLAGS_TILE_MAP_TILE_SET]) {
+			// Update all cells.
+			for (KeyValue<Vector2i, CellData> &kv : tile_map) {
+				_physics_update_cell(kv.value);
+			}
+		} else {
+			// Update dirty cells.
+			for (SelfList<CellData> *cell_data_list_element = dirty.cell_list.first(); cell_data_list_element; cell_data_list_element = cell_data_list_element->next()) {
+				CellData &cell_data = *cell_data_list_element->self();
+				_physics_update_cell(cell_data);
+			}
+		}
 
-	PhysicsServer2D *ps = PhysicsServer2D::get_singleton();
-	RID space = tile_map_node->get_world_2d()->get_space();
+		Transform2D gl_transform = tile_map_node->get_global_transform();
+		if (dirty.flags[DIRTY_FLAGS_TILE_MAP_LOCAL_XFORM] || dirty.flags[DIRTY_FLAGS_TILE_MAP_XFORM]) {
+			bool in_editor = false;
+#ifdef TOOLS_ENABLED
+			in_editor = Engine::get_singleton()->is_editor_hint();
+#endif
+			if (!tile_map_node->is_collision_animatable() || in_editor) {
+				for (KeyValue<Vector2i, CellData> &kv : tile_map) {
+					const CellData &cell_data = kv.value;
+
+					for (RID body : cell_data.bodies) {
+						if (body.is_valid()) {
+							Transform2D xform(0, tile_map_node->map_to_local(bodies_coords[body]));
+							xform = gl_transform * xform;
+							PhysicsServer2D::get_singleton()->body_set_state(body, PhysicsServer2D::BODY_STATE_TRANSFORM, xform);
+						}
+					}
+				}
+			}
+		}
+	}
 
-	SelfList<TileMapQuadrant> *q_list_element = r_dirty_quadrant_list.first();
-	while (q_list_element) {
-		TileMapQuadrant &q = *q_list_element->self();
+	// -----------
+	// Mark the physics state as up to date.
+	_physics_was_cleaned_up = forced_cleanup;
+}
 
-		// Clear bodies.
-		for (RID body : q.bodies) {
+void TileMapLayer::_physics_clear_cell(CellData &r_cell_data) {
+	PhysicsServer2D *ps = PhysicsServer2D::get_singleton();
+
+	// Clear bodies.
+	for (RID body : r_cell_data.bodies) {
+		if (body.is_valid()) {
 			bodies_coords.erase(body);
 			ps->free(body);
 		}
-		q.bodies.clear();
+	}
+	r_cell_data.bodies.clear();
+}
+
+void TileMapLayer::_physics_update_cell(CellData &r_cell_data) {
+	const Ref<TileSet> &tile_set = tile_map_node->get_tileset();
+	Transform2D gl_transform = tile_map_node->get_global_transform();
+	RID space = tile_map_node->get_world_2d()->get_space();
+	PhysicsServer2D *ps = PhysicsServer2D::get_singleton();
 
-		// Recreate bodies and shapes.
-		for (const Vector2i &E_cell : q.cells) {
-			TileMapCell c = get_cell(E_cell, true);
+	// Recreate bodies and shapes.
+	TileMapCell c = r_cell_data.cell;
```

Comment:
The TileMapCells are small enough that it doesn't matter if you pass them by reference or by value, but seems like you are doing it inconsistently.